### PR TITLE
Adds `__hash__` method to `@dataclasses.dataclass`, refs #11463

### LINKED ISSUE #11463: `dataclass` is not `Hashable`
```py
from dataclasses import dataclass
from typing import Hashable


@dataclass
class SusDataclass:
    b = "AMONGUS"


h: Hashable = SusDataclass()  # SUS ALERT!

d = {
    SusDataclass(): 1  # SUS ALERT! fails at runtime
}
```

### PR DESCRIPTION
**Pull request by sobolevn**
Closes #11463
Closes https://github.com/python/mypy/issues/11495

I still need to add a test case for `3.9`, where there's no `unsafe_hash` arg, but `__hash__` is still added somehow.
Right now I just want to see the CI results.

### CODE REVIEW

File: test-data/unit/check-dataclasses.test [hunk 1]
```diff
@@ -1420,4 +1425,130 @@ class Some:
...
+
+@dataclass(unsafe_hash=False, eq=False, frozen=False)
+class FirstCase1:
+    x: int
+reveal_type(FirstCase1(1).__hash__)  # N: Revealed type is "def () -> builtins.int"
+
+@dataclass(unsafe_hash=False, eq=False, frozen=False)
+class FirstCase2:
+    x: int
+reveal_type(FirstCase2(1).__hash__)  # N: Revealed type is "def () -> builtins.int"
```
Comment (JukkaL) [inline test-data/unit/check-dataclasses.test:1447]: This looks identical to `FirstCase1`. Is this redundant or should this be different?

File: mypy/plugins/dataclasses.py [hunk 2]
```diff
@@ -231,6 +241,67 @@ def add_slots(self,
...
+        else:
+            # "Has no __eq__, but hashable" or
+            # "Not frozen, but hashable" or
+            # "Frozen, so hashable"
+            if existing:
+                # When class already has `__hash__` defined, we do not allow
+                # to override it. So, raise an error and do nothing.
+                self._ctx.api.fail(
+                    'Cannot overwrite attribute "__hash__" in class "{}"'.format(
+                        self._ctx.cls.name,
```
Comment (JukkaL) [inline mypy/plugins/dataclasses.py:291]: Nit: move the error message to message registry?

**Review by JukkaL — COMMENTED** [review]
Thanks for the PR! I like that all the possible cases are carefully tested. Left some minor comments, overall looks good.

File: test-data/unit/check-dataclasses.test (hunk 1 above)
Comment (sobolevn) [reply test-data/unit/check-dataclasses.test:1447]: Indeed it was. Fixed! I've also found a bunch of missed cases.

File: mypy/plugins/dataclasses.py [hunk 3]
```diff
@@ -231,6 +241,67 @@ def add_slots(self,
 
         info.slots = {attr.name for attr in attributes}
 
+    def add_hash(self,
+                 info: TypeInfo,
+                 decorator_arguments: Dict[str, bool]) -> None:
+        unsafe_hash = decorator_arguments.get('unsafe_hash', False)
+        eq = decorator_arguments['eq']
+        frozen = decorator_arguments['frozen']
+        cond = (unsafe_hash, eq, frozen)
```
[the remark this answers is not in the document]
Comment (sobolevn) [reply mypy/plugins/dataclasses.py:250]:
```python
if ((not unsafe_hash and not eq and not frozen)
            or (not unsafe_hash and not eq and frozen)):
```

I think that both are quite hard to read 😞 
Mine was inspired by pattern matching technique I use in different functional languages.

Anyways, I've changed it! 👍

**Comment by JukkaL** [comment]
The mypy_primer failure seems to indicate a real issue. For example, this generates a false positive:
```py
from dataclasses import dataclass

a = f()  # Forward ref to force two semantic analysis passes

@dataclass(unsafe_hash=True)
class C:  # Cannot overwrite attribute "__hash__" in class "C"
    x: str

def f(): pass
```

It looks like the new logic doesn't work correctly if we need multiple semantic analysis passes. If the previous pass had added a `__hash__` method, we shouldn't complain about it in the second pass.

**Comment by sobolevn** [comment]
Let's try this solution. Now I check that given variable was not plugin generated.

**Comment by JukkaL** [comment]
Can you fix the conflicts? I can do another round of review once the PR is up-to-date.

**Comment by sobolevn** [comment]
I somehow nuked my old fork and that's why all my old PRs are now closed.
Thanks a lot to @AlexWaygood for letting me know.

I can recreate some of them. Here's a list of things I still want to get merged:
- https://github.com/python/mypy/pull/11496
- https://github.com/python/mypy/pull/11790
- https://github.com/python/mypy/pull/11793
- https://github.com/python/mypy/pull/11693
- https://github.com/python/mypy/pull/12088

**Comment by AlexWaygood** [comment]
> I can recreate some of them. Here's a list of things I still want to get merged:

I was a big fan of these, as well :)

- #11885
- #11992

**Pull request closed without merging**